src/compiler/checker.ts

```diff
@@ -4722,6 +4846,1104 @@ namespace ts {
+                    }
+                    return statements;
+                }
+
+                function canHaveExportModifier(node: Statement) {
+                    return isEnumDeclaration(node) ||
+                            isVariableStatement(node) ||
+                            isFunctionDeclaration(node) ||
+                            isClassDeclaration(node) ||
+                            (isModuleDeclaration(node) && !isExternalModuleAugmentation(node) && !isGlobalScopeAugmentation(node)) ||
+                            isInterfaceDeclaration(node) ||
+                            isTypeDeclaration(node);
+                }
+
+                function addExportModifier(statement: Statement) {
+                    const flags = (getModifierFlags(statement) | ModifierFlags.Export) & ~ModifierFlags.Ambient;
+                    statement.modifiers = createNodeArray(createModifiersFromModifierFlags(flags));
+                    statement.modifierFlagsCache = 0;
+                }
+
+                function visitSymbolTable(symbolTable: SymbolTable, suppressNewPrivateContext?: boolean, propertyAsAlias?: boolean) {
+                    const oldDeferredList = deferredPrivates;
+                    if (!suppressNewPrivateContext) {
+                        deferredPrivates = createMap();
+                    }
+                    symbolTable.forEach((symbol: Symbol) => {
+                        serializeSymbol(symbol, /*isPrivate*/ false, !!propertyAsAlias);
+                    });
+                    if (!suppressNewPrivateContext) {
+                        // deferredPrivates will be filled up by visiting the symbol table
+                        // And will continue to iterate as elements are added while visited `deferredPrivates`
+                        // (As that's how a map iterator is defined to work)
+                        deferredPrivates!.forEach((symbol: Symbol) => {
+                            serializeSymbol(symbol, /*isPrivate*/ true, !!propertyAsAlias);
+                        });
+                    }
+                    deferredPrivates = oldDeferredList;
+                }
+
+                function includePrivateSymbol(symbol: Symbol) {
+                    Debug.assertDefined(deferredPrivates);
+                    deferredPrivates!.set("" + getSymbolId(symbol), symbol);
+                }
+
+                // Appends a `declare` and/or `export` modifier if the context requires it, and then adds `node` to `result` and returns `node`
+                // Note: This _mutates_ `node` without using `updateNode` - the assumption being that all nodes should be manufactured fresh by the node builder
+                function addResult(node: Statement, additionalModifierFlags: ModifierFlags) {
+                    let newModifierFlags: ModifierFlags = ModifierFlags.None;
+                    if (((additionalModifierFlags & ModifierFlags.Export) && enclosingDeclaration &&

```

Comment:
this condition is so complex I'd recommend breaking out the inner part at least into a temp. Maybe more obvious formatting can save it as-is, but I don't think so.